presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -399,27 +418,136 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return distinctingGroupedAccumulator.getEstimatedSize() + intermediateState.sizeOf();
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return distinctingGroupedAccumulator.getFinalType();
+        }
+
+        @Override
+        public Type getIntermediateType()
+        {
+            return new ArrayType(RowType.anonymous(spillingTypes));
+        }
+
+        @Override
+        public void addInput(GroupByIdBlock groupIds, Page page)
+        {
+            for (int pos = 0; pos < groupIds.getPositionCount(); pos++) {
+                long groupId = groupIds.getGroupId(pos);
+                intermediateState.ensureCapacity(groupId);
+                RowBlockBuilder rowBlockBuilder = intermediateState.get(groupId);
+                if (rowBlockBuilder == null) {
+                    rowBlockBuilder = new RowBlockBuilder(spillingTypes, null, (int) groupIds.getGroupCount());
+                }
+
+                BlockBuilder output = rowBlockBuilder.beginBlockEntry();
+                for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                    spillingTypes.get(channel).appendTo(page.getBlock(channel), pos, output);
+                }
+                rowBlockBuilder.closeEntry();
+
+                intermediateState.set(groupId, rowBlockBuilder);
+            }
+            // TODO do this in prepareFinal() instead because this duplicates data in intermediateState and distincting hash table
+            distinctingGroupedAccumulator.addInput(groupIds, page);
+        }
+
         @Override
         public void addIntermediate(GroupByIdBlock groupIdsBlock, Block block)
         {
-            throw new UnsupportedOperationException();
+            ArrayBlock arrayBlock = (ArrayBlock) block;
+            int newPositionCount = arrayBlock.getRawElementBlock().getPositionCount();
+            long[] newGroupIds = new long[newPositionCount];
+            boolean[] nulls = new boolean[newPositionCount];
+            int currentRowBlockIndex = 0;
+            for (int groupIdPosition = 0; groupIdPosition < groupIdsBlock.getPositionCount(); groupIdPosition++) {
+                for (int unused = 0; unused < arrayBlock.getBlock(groupIdPosition).getPositionCount(); unused++) {
+                    newGroupIds[currentRowBlockIndex] = groupIdsBlock.getGroupId(groupIdPosition);
+                    nulls[currentRowBlockIndex] = groupIdsBlock.isNull(groupIdPosition);
+                    currentRowBlockIndex++;
+                }
+            }
+
+            Block[] blocks = new Block[spillingTypes.size() + 1];
+            for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                blocks[channel + 1] = ((RowBlock) arrayBlock.getRawElementBlock()).getRawFieldBlocks()[channel]; // super ugly exposing block internals
```

Comment:
Block internals need to be exposed for this implementation. Let me know if there are any alternatives.